src/aggregation/bucket/filter.rs

```diff
@@ -0,0 +1,380 @@
+                     serialization support.",
+                ))
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for FilterAggregation {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where D: Deserializer<'de> {
+        // Deserialize as query string
+        let query_string = String::deserialize(deserializer)?;
+        Ok(FilterAggregation::new(query_string))
+    }
+}
+
+// PartialEq is required because AggregationVariants derives it
+// We implement it manually to handle Box<dyn Query> which doesn't impl PartialEq
+impl PartialEq for FilterAggregation {
+    fn eq(&self, other: &Self) -> bool {
+        match (&self.query, &other.query) {
+            (FilterQuery::QueryString(a), FilterQuery::QueryString(b)) => a == b,
+            // Direct queries cannot be compared for equality
+            _ => false,
+        }
+    }
+}
+
+/// Document evaluator for filter queries
+/// This avoids running separate query executions and instead evaluates queries per document
+struct DocumentQueryEvaluator {
+    /// Cached weight for the current segment
+    weight: Option<Box<dyn Weight>>,
+    /// Note: We can't pre-create the scorer because it maintains state (current doc position)
+    /// that would be invalid across multiple document evaluations.
+    /// Instead, we cache the SegmentReader which is cheap (Arc-based internally).
+    segment_reader: Option<SegmentReader>,
+}
+
+impl DocumentQueryEvaluator {
+    /// Create and initialize a document query evaluator for a segment
+    ///
+    /// Note: SegmentReader is cloned here but the clone is cheap because:
+    /// - SegmentReader uses Arc<OnceLock<...>> for all data structures
+    /// - All data is memory-mapped, so no actual data is copied
+    /// - Clone just increments reference counts
+    fn new(
+        query: Box<dyn Query>,
+        schema: Schema,
+        segment_reader: &SegmentReader,
+    ) -> crate::Result<Self> {
+        use crate::query::EnableScoring;
+        let weight = Some(query.weight(EnableScoring::disabled_from_schema(&schema))?);
+        Ok(Self {
+            weight,
+            segment_reader: Some(segment_reader.clone()),
+        })
+    }
+
+    /// Evaluate if a document matches the filter query
+    /// This is the core performance-critical method
+    pub fn matches_document(&self, doc: DocId) -> crate::Result<bool> {
+        let weight = self.weight.as_ref().ok_or_else(|| {
+            TantivyError::InvalidArgument(
+                "DocumentQueryEvaluator not initialized for segment".to_string(),
+            )
+        })?;
+
+        let segment_reader = self.segment_reader.as_ref().ok_or_else(|| {
+            TantivyError::InvalidArgument(
+                "DocumentQueryEvaluator not initialized for segment".to_string(),
+            )
+        })?;
+
+        // This already handles all optimizations (fast fields, bitsets, etc.)
+        let mut scorer = weight.scorer(segment_reader, 1.0)?;
```

Comment:
I think this would be way too slow for aggregations that have many hits. I think a better way would be to execute the search and collect it in a BitSet upfront (to be able to handle out of order docs due to caching), same as we do for range queries. It's not ideal though in cases we skip a lot, since we may fill the Bitset and never read it.

Range queries on the inverted index would very likely overload the system https://github.com/quickwit-oss/tantivy/blob/main/src/query/range_query/range_query.rs#L215

I'm currently refactoring aggregations to be able to store precomputed data in a better way